iodata/formats/wfx.py

```diff
@@ -18,224 +18,156 @@
+        '<Number of Core Electrons>': 'num_core_electrons',
+        '<Number of Alpha Electrons>': 'num_alpha_electron',
+        '<Number of Beta Electrons>': 'num_beta_electron',
+        '<Number of Primitives>': 'num_primitives',
+        '<Electronic Spin Multiplicity>': 'spin_multi',
+        # float numbers
+        '<Net Charge>': 'charge',
+        '<Energy = T + Vne + Vee + Vnn>': 'energy',
+        '<Virial Ratio (-V/T)>': 'virial_ratio',
+        '<Nuclear Virial of Energy-Gradient-Based Forces on Nuclei, W>': 'nuc_viral',
+        '<Full Virial Ratio, -(V - W)/T>': 'full_virial_ratio',
+    }
+    labels_array = {
+        '<Nuclear Cartesian Coordinates>': 'atcoords',
+        '<Nuclear Charges>': 'nuclear_charge',
+        '<Atomic Numbers>': 'atnums',
+        '<Primitive Exponents>': 'exponents',
+        '<Primitive Centers>': 'centers',
+        '<Primitive Types>': 'types',
+        '<Molecular Orbital Energies>': 'mo_energy',
+        '<Molecular Orbital Occupation Numbers>': 'mo_occ',
+        '<Molecular Orbital Primitive Coefficients>': 'mo_coeff',
+    }
+    labels_other = {
+        '<Nuclear Names>': 'nuclear_names',
+        '<Molecular Orbital Spin Types>': 'mo_spin',
+        '<Nuclear Cartesian Energy Gradients>': 'nuclear_gradient',
     }
 
-    dict_int = {}
-    for key, val in int_label.items():
-        int_info = _helper_section(f_content=f_content,
-                                   start=val[0],
-                                   end=val[1])
-        if int_info:
-            dict_int[key] = np.array(int_info[0], dtype=np.int)
+    # load raw data
+    data = parse_wfx(lit)
+    # process raw data
+    result = {}
+    for key, value in data.items():
+        if key in labels_str:
+            result[labels_str[key]] = value[0]
+        elif key in labels_number:
+            result[labels_number[key]] = np.array(eval(value[0]))
+        elif key in labels_array:
+            breakdown = [eval(item) for line in value for item in line.split()]
+            result[labels_array[key]] = np.array(breakdown)
+        elif key in labels_other:
+            result[labels_other[key]] = value
         else:
-            dict_int[key] = np.array(None)
-
-    return dict_int
+            warnings.warn("Not recognized label, skip {0}".format(key))
 
+    # reshape some arrays
+    result['atcoords'] = result['atcoords'].reshape(-1, 3)
+    result['mo_coeff'] = result['mo_coeff'].reshape(result['num_primitives'], -1, order='F')
+    # process mo spin type
+    mo_spin_list = [i.split() for i in result['mo_spin']]
+    mo_spin_type = np.array(mo_spin_list, dtype=np.unicode_).reshape(-1, 1)
+    result['mo_spin'] = mo_spin_type[mo_spin_type[:, 0] != 'and']
+    # process nuclear gradient
+    gradient_mix = np.array([i.split() for i in result.pop('nuclear_gradient')]).reshape(-1, 4)
+    result['gradient_atoms'] = gradient_mix[:, 0].astype(np.unicode_)
+    result['gradient'] = gradient_mix[:, 1:].astype(float)
```

Comment:
`gradient_atoms` does not get used and I'm not sure why it is present in the file format? Can it happen that nuclear forces are absent for some atoms? If not, the above three lines can be replaced by

```python
result['gradient'] = np.array([line.split()[1:] for line in result.pop('nuclear_gradient')], dtype=float)
```
(shape should be ok)

Reply:
The nuclear gradient in WFX is composed of gradient atoms (`gradient_atoms`) and the corresponding gradient (`gradient`). Because not all nuclear have gradient data, we need to keep them. I tried to make them a dictionary,
```python
    result = {
        'title': data['title'],
        'atcoords': data['atcoords'],
        'atnums': data['atnums'],
        'obasis': obasis,
        'mo': mo,
        'atgradient': {'gradient_atoms': data['gradient_atoms'], 'gradient': data['gradient']},
    }
```
Will this be OK?

Reply:
That will not work because `atgradient` has to be an `(natom,3)` array. Also adding new undocumented attributes, like `gradient` will not work. We have such restrictions in place to make sure that data loaded from different files can be used in the same way, independently of the file format. I would suggest the following:
- Remove `gradient` from the dictionary
- Turn `atgradient` into a numpy array with shape `(natom, 3)` and fill in as much elements as possible, leaving undefined ones `NaN`.

Reply:
Sorry, I misread your code. Let my try to clarify. The following would be good:

```python
result = {
        # some lines omitted
        'atgradient': atgradient,
        # some lines omitted
    }
```

where `atgradient` is simply a numpy array with shape `(natom, 3)`, where as much as possible values are filled in, leaving the rest to `NaN`

Reply:
Thank you for the clarification and Farnaz's codes works.